### DISCUSSION TITLE
graphCache not working as expected

### DISCUSSION THREAD

**Discussion by bubbaspaarx** [urql-graphql/urql#1441] [Q&A] [answered]
We are trying to move over to graphCache from the default document-cache in order to reduce the amount of fetches caused when we have mutations. 

What we are trying to establish right now, is if it will actually fix this issue or if there's an alternative way.

At the moment we have a list of authors with a list of their respective books and inside each book are the pages. We make a query for the list of authors, then a separate query for the book and pages per author which all sit inside a separate component. 

If we delete a page from a book, all books for all authors appear to be refetched and the book components re-rendered. This we understood to be a constraint of document-caching.  However, we are now attempting to resolve this by using graph-cache and invalidating the individual books so only theirs are refetched. So far we still appear to be re-rendering all book components for all authors and refetching them. Is this normal or have we made a mistake with the cache.invalidation?

our cache invalidation 

```
updates: {
        Mutation: {
          deletePage: (result, args, cache) => {
            cache.invalidate({
              __typename: 'Book',
              id: result.deletePage.book.id
            })
          }
        }
      }

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

This is a side-effect of not knowing what the author.books array contains as a type, let's look at how this works.

When we invalidate the book this book becomes `['Book:1', null]` as we have invalidated `Book:2`, if your author.books type is `[Book]!` we can safely render the partial data (requires schema-awareness) if it's `[Book!]!` we can't safely render partial data.

So invalidate won't work for this particular case, a better solution for this specific case would be to get the affected entity in question and update its array with updateQuery.

**Reply by bubbaspaarx** [reply]
All of our types are the latter [Book!]!

are you saying that we should create a cache.updateQuery within the cacheExchange? We have put it together below but it doesn't appear to be working. Also, does this mean we don't need schema awareness too?

```
updates: {
        Mutation: {
          deletePage: (result, args, cache) => {
            cache.updateQuery(
              {
                query: getBooksQuery,
                variables: {
                  authorId:
                    result.deletePage.book.author.id
                }
              },
              (data) => data
            )
          }
        }
      }
```

**Reply by kitten** [reply]
Have you seen the examples ont he "Cache Updates" page? They also include deletions https://formidable.com/open-source/urql/docs/graphcache/cache-updates/

What's important to note here is even if you invalidate an item, all queries associated to that item must refetch. The only way to prevent refetches is to 100% provide the cache enough information to resolve all queries from its cache with your changes, which sometimes means that you have to write an updater (like you're doing) to reflect the changes that a mutation has made.

**Reply by bubbaspaarx** [reply]
This is where I'm a little confused. I'm actually trying to do the opposite i WANT to refetch the query because I need the backend to provide additional information that I can't get without this operation. If this is the case, do I simply chain a query on to the delete mutation? I feel like i've over complicated things.

**Reply by kitten** [reply]
If you're deleting for instance your `Book` and you just want to refetch then `cache.invalidate` is indeed the right method to use. What's important here is that it'll refetch all queries/data that contains this exact entity though

**Reply by bubbaspaarx** [reply]
Ok, i think I understand what you mean. When you say entity, do you mean Book or Book:1? Ideally, I only want to refetch information for Book:1 not all Books. Is this even possible?

**Reply by kitten** [reply]
Yes, that's exactly what's happening. So when you invalidate `{ __typename: 'Book', id }` then only that exact entity is invalidated.

**Reply by bubbaspaarx** [reply]
Just wanted to say thank you for your help and understanding and for URQL too. It's a brilliant tool. I'm still having a small issue but think I understand it more now. Much appreciated

**Reply by bubbaspaarx** [reply]
So we've made some progress with the pages situation. adding and deleting pages now refetch their parent book queries as expected. We are now seeing a slightly different issue though when creating a new book. In this situation, we already use the 'addtionalTypenames' context so that empty arrays of results don't prevent the fetch for newly added books (this works fine under default cache). However, the situation we are seeing seems to disregard this and also doesn't refetch even if other books are in the array. 
We assumed that the original query we make for all books per author would be triggered when a new book is added. We even added the 'author' as an additionalTypename the mutation but this doesn't make a difference either. What are we missing here? Many thanks

**Reply by kitten** [reply]
The `additionalTypenames` API isn't used by Graphcache since it doesn't operate on heuristics but on a normalised cache.

If you're just looking to invalidate again with Graphcache, as before, you can also invalidate fields, eg a listing field on `Query`

**Reply by JoviDeCroock** [reply]
`additionalTypenames` is a concern for the document-based cache, when we want to add items to an entity we'll need to write an updater. So a general mindset to take is "is the entity already available in cache" if it is in cache we'll automatically update it. "does the respond signal us the updated entity", this means that if you'd return the parent entity with the updated array graphCache can automatically update the entity in-memory. If you only return the newly created entity you'll have to provide the logic to append it to the array.

**Reply by bubbaspaarx** [reply]
Thank you again for your help with this. We've stuck with the graphcache and removed the additionalTypename fields too. To select the correct fields to invalidate, we tried invalidating the author, but as the original query for author was for all authors, it ended up overfetching. We then used the field invalidation technique which initially (using the example on the docs) ended up invalidating all books, we could narrow it down to the fetch we needed by doing this

```
          createBook: (result, _args, cache) => {
            const key = 'Query'
            cache.inspectFields(key).filter((field) =>
                  field.fieldKey === `books({"authorId":"${result.createBook.author.id}"})`)
              .forEach((field) => {
                cache.invalidate(key, field.fieldKey)
              })
          }
```
